srml/babe/src/lib.rs

```diff
@@ -191,14 +212,48 @@ decl_storage! {
 			storage: &mut (sr_primitives::StorageOverlay, sr_primitives::ChildrenStorageOverlay),
 			config: &GenesisConfig
 		| {
-			runtime_io::with_storage(
+			sr_io::with_storage(
 				storage,
 				|| Module::<T>::initialize_authorities(&config.authorities),
 			);
 		})
 	}
 }
 
+fn equivocation_is_valid<T: Trait>(
+	equivocation: &EquivocationProof<T::Header, AuthorityId, AuthoritySignature>,
+) -> bool {
+	let first_header = &equivocation.first_header;
+	let second_header = &equivocation.second_header;
+
+	if first_header == second_header {
+		return false
+	}
+
+	let maybe_first_slot = get_slot::<T::Header>(&first_header);
+	let maybe_second_slot = get_slot::<T::Header>(&second_header);
+
+	if maybe_first_slot.is_err() || maybe_second_slot.is_err() {
+		return false
+	}
+	let first_slot = maybe_first_slot.expect("checked before; qed");
+	let second_slot = maybe_second_slot.expect("checked before; qed");
+
+	if equivocation.slot == first_slot && first_slot == second_slot {
+		let author = &equivocation.identity;
+
+		if !author.verify(&first_header.hash(), &equivocation.first_signature) {
```

Comment:
this is a serious misunderstanding of BABE - the hash includes the seal includes the signature, so it will always fail. I'm not sure why the entire header needs to go into the proof at all.